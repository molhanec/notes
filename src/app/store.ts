import { Action, configureStore, getDefaultMiddleware, ThunkAction } from '@reduxjs/toolkit'
import { connectRouter, routerMiddleware } from 'connected-react-router'
import { createBrowserHistory } from 'history'

import counterReducer from '../features/counter/counterSlice'
import notesReducer from '../features/notes/notesSlice'

export const history = createBrowserHistory()

export const store = configureStore({
  reducer: {
    counter: counterReducer,
    router: connectRouter(history) as any,
    notes: notesReducer,
  },
  middleware: [...getDefaultMiddleware(), routerMiddleware(history)],
})

export type RootState = ReturnType<typeof store.getState>
export type AppThunk = ThunkAction<void, RootState, unknown, Action<string>>
