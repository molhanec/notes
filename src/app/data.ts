import { AppThunk } from "./store"
import { loadNotesAsync } from "../features/noteList/notesSlice"
import { setProgress, setSuccess, setError } from "../features/info/infoSlice"

const rootDataUrl =
  "https://private-anon-3ab0da907c-note10.apiary-mock.com/notes"

const checkResponseStatus = (response: Response) => {
  if (response.ok) {
    return response
  } else {
    return Promise.reject(
      `Error fetching data: ${response.status} ${response.statusText}`
    )
  }
}

export const fetchJson = <T>(path: string) =>
  fetch(rootDataUrl + path)
    .then(checkResponseStatus)
    .then(response => response!.json() as Promise<T>)

export const addNoteAsync = (title: string): AppThunk => async dispatch => {
  dispatch(setProgress())
  try {
    fetch(rootDataUrl, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        title,
      }),
    }).then(checkResponseStatus)

    dispatch(setSuccess("New note added"))
    dispatch(loadNotesAsync())
  } catch {
    dispatch(setError("Cannot add new note"))
  }
}
