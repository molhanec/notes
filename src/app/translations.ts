import { useSelector } from "react-redux"
import { RootState } from "./store"

export type LanguageCodes = "cz" | "en"
export const defaultLanguage = "cz"

interface Phrases {
  appTitle: string
  noteLoadFailure: string
  notesLoadFailure: string

  noteCreateSuccess: string
  noteCreateFailure: string
  noteUpdateSuccess: string
  noteUpdateFailure: string
  noteDeleteSuccess: string
  noteDeleteFailure: string

  formTitle: string
  formSave: string
  formDelete: string

  editFormHeader: string
  newFormHeader: string
}

export type PhraseKeys = keyof Phrases

const enPhrases: Phrases = {
  appTitle: "Notes",
  noteLoadFailure: "Cannot load note",
  notesLoadFailure: "Cannot load notes",

  noteCreateSuccess: "New note added",
  noteCreateFailure: "Cannot add new note",
  noteUpdateSuccess: "Note updated",
  noteUpdateFailure: "Cannot update note",
  noteDeleteSuccess: "Note deleted",
  noteDeleteFailure: "Cannot delete note",

  formTitle: "Title",
  formSave: "Save",
  formDelete: "Delete",

  editFormHeader: "Note details",
  newFormHeader: "Create new note",
}

const czPhrases: Phrases = {
  appTitle: "Poznámky",
  noteLoadFailure: "Nelze načíst poznámku",
  notesLoadFailure: "Nelze načíst poznámky",

  noteCreateSuccess: "Přidána poznámka",
  noteCreateFailure: "Nelze přidat poznámku",
  noteUpdateSuccess: "Poznámka upravena",
  noteUpdateFailure: "Nelze upravit poznámku",
  noteDeleteSuccess: "Poznámka smazána",
  noteDeleteFailure: "Nelze smazat poznámku",

  formTitle: "Nadpis",
  formSave: "Uložit",
  formDelete: "Smazat",

  editFormHeader: "Detaily poznámky",
  newFormHeader: "Vytvořit novou poznámku",
}

export const getPhrase = (language: "cz" | "en", key: PhraseKeys): string => {
  switch (language) {
    case "cz":
      return czPhrases[key]
    case "en":
      return enPhrases[key]
  }
}

export const useTranslate = () => {
  const language = useSelector((state: RootState) => state.language.code)
  return (key: PhraseKeys) => getPhrase(language, key)
}
