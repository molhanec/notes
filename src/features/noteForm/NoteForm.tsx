import React, { FormEvent, useState, ChangeEvent } from "react"
import { connect, ResolveThunks } from "react-redux"
import { addNoteAsync } from "../../app/data"

type Props = ResolveThunks<typeof mapDispatchToProps>

export const NoteForm = ({ addNoteAsync }: Props) => {
  const [title, setTitle] = useState<string>("")

  const titleChange = (event: ChangeEvent<HTMLInputElement>) => {
    setTitle(event.target.value)
  }

  const submit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    if (title.trim()) addNoteAsync(title.trim())
    setTitle("")
  }

  return (
    <form onSubmit={submit}>
      <label htmlFor="title">Title </label>
      <input id="title" value={title} onChange={titleChange} />
      <button disabled={!title.trim()}>Save</button>
    </form>
  )
}

const mapDispatchToProps = {
  addNoteAsync,
}

const ConnectedNoteForm = connect(null, mapDispatchToProps)(NoteForm)
export default ConnectedNoteForm
