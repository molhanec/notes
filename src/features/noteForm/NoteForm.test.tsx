import React from "react"
import { render, fireEvent } from "@testing-library/react"
import { Provider } from "react-redux"
import { store } from "../../app/store"
import { NoteForm } from "./NoteForm"
import { getPhrase } from "../../app/translations"

describe("<NoteForm>", () => {
  test("matches snapshot", () => {
    const { container } = render(
      <Provider store={store}>
        <NoteForm
          formHeader="Header"
          initialTitle="initial"
          onSubmit={jest.fn()}
          onDelete={jest.fn()}
        />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })

  test("calls onSubmit when Save button clicked", () => {
    const onSubmit = jest.fn()
    const { getByText } = render(
      <Provider store={store}>
        <NoteForm
          formHeader="Header"
          initialTitle="initial"
          onSubmit={onSubmit}
          onDelete={jest.fn()}
        />
      </Provider>
    )

    const saveButton = getByText(getPhrase("cz", "formSave"))
    fireEvent.click(saveButton)

    expect(onSubmit).toBeCalled()
  })

  test("calls onDelete when Delete button clicked", () => {
    const onDelete = jest.fn()
    const { getByText } = render(
      <Provider store={store}>
        <NoteForm
          formHeader="Header"
          initialTitle="initial"
          onSubmit={jest.fn()}
          onDelete={onDelete}
        />
      </Provider>
    )

    const deleteButton = getByText(getPhrase("cz", "formDelete"))
    fireEvent.click(deleteButton)

    expect(onDelete).toBeCalled()
  })
})
