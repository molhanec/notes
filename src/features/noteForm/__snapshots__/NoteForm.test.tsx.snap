// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NewNoteForm> matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Vytvořit novou poznámku
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
      >
        <label
          class=""
          for="title"
        >
          Nadpis
        </label>
        <input
          class="form-control"
          id="title"
          type="text"
          value=""
        />
        <button
          class="mt-3 btn btn-primary disabled"
          disabled=""
        >
          Uložit
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`<NoteForm> matches snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      Header
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
      >
        <label
          class=""
          for="title"
        >
          Nadpis
        </label>
        <input
          class="form-control"
          id="title"
          type="text"
          value="initial"
        />
        <button
          class="mt-3 btn btn-primary"
        >
          Uložit
        </button>
        <button
          class="mt-3 ml-1 btn btn-danger"
          type="button"
        >
          Smazat
        </button>
      </form>
    </div>
  </div>
</div>
`;
